File: cpp/problems/leetcode/0045_jump_game_ii/solution.cpp
// LeetCode 45
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int jump(vector<int>& nums) {
        int steps = 0, end = 0, far = 0;
        for (int i = 0; i < (int)nums.size() - 1; ++i) {
            far = max(far, i + nums[i]);
            if (i == end) {
                ++steps;
                end = far;
            }
        }
        return steps;
    }
};

int main() {
    vector<int> a{2, 3, 1, 1, 4};
    assert(Solution().jump(a) == 2);
    vector<int> b{2, 3, 0, 1, 4};
    assert(Solution().jump(b) == 2);
    vector<int> c{0};
    assert(Solution().jump(c) == 0);
    cout << "0045 OK" << endl;
    return 0;
}

File: python/problems/leetcode/0045_jump_game_ii/solution.py
"""LeetCode 45: Jump Game II."""

from __future__ import annotations


class Solution:
    def jump(self, nums: list[int]) -> int:
        steps = end = far = 0
        for i in range(len(nums) - 1):
            far = max(far, i + nums[i])
            if i == end:
                steps += 1
                end = far
        return steps


if __name__ == "__main__":
    assert Solution().jump([2, 3, 1, 1, 4]) == 2
    assert Solution().jump([2, 3, 0, 1, 4]) == 2
    assert Solution().jump([0]) == 0
    print("0045 OK")
